core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java

```diff
@@ -616,6 +613,65 @@ public EnvelopePart copyAndShift(double positionDelta, double minPosition, doubl
                 new HashMap<>(attrs), newPositions.toArray(), newSpeeds.toArray(), newTimeDeltas.toArray());
     }
 
+    /**
+     * Compute the minimum EnvelopePart between 2 envelope parts with an intersecting range [a, b].
+     * The resulting envelope part starts at the envelope with the minimum speed at a and ends at b.
+     */
+    public static EnvelopePart min(
+            EnvelopePart envelopePartA, EnvelopePart envelopePartB, Iterable<SelfTypeHolder> attrs) {
+        var beginPosA = envelopePartA.getBeginPos();
+        var beginPosB = envelopePartB.getBeginPos();
+        var endPosA = envelopePartA.getEndPos();
+        var endPosB = envelopePartB.getEndPos();
+        assert (beginPosA < endPosB && endPosA > beginPosB);
+        var startIntersectingRange = Math.max(beginPosA, beginPosB);
+        var start = envelopePartA.interpolateSpeed(startIntersectingRange)
+                        <= envelopePartB.interpolateSpeed(startIntersectingRange)
+                ? beginPosA
+                : beginPosB;
+        var end = Math.min(endPosA, endPosB);
+
+        TreeSet<Double> keyPositions =
+                Arrays.stream(envelopePartA.positions).boxed().collect(Collectors.toCollection(TreeSet::new));
+        keyPositions.addAll(Arrays.stream(envelopePartB.positions).boxed().collect(Collectors.toSet()));
+        keyPositions = new TreeSet<>(keyPositions.subSet(start, true, end, true));
```

Comment:
```suggestion
        NavigableSet<Double> keyPositions =
                Arrays.stream(envelopePartA.positions).boxed().collect(Collectors.toCollection(TreeSet::new));
        keyPositions.addAll(Arrays.stream(envelopePartB.positions).boxed().collect(Collectors.toSet()));
        keyPositions = keyPositions.subSet(start, true, end, true);
```

Avoids instantiating a new TreeSet from the subset